File: app/src/main/java/aykuttasil/com/modernapp/util/AutoClearedValue.java
package aykuttasil.com.modernapp.util;


import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;

/**
 * A value holder that automatically clears the reference if the Fragment's view is destroyed.
 *
 * @param <T>
 */
public class AutoClearedValue<T> {
    private T value;

    public AutoClearedValue(Fragment fragment, T value) {
        FragmentManager fragmentManager = fragment.getFragmentManager();
        fragmentManager.registerFragmentLifecycleCallbacks(
                new FragmentManager.FragmentLifecycleCallbacks() {
                    @Override
                    public void onFragmentViewDestroyed(FragmentManager fm, Fragment f) {
                        if (f == fragment) {
                            AutoClearedValue.this.value = null;
                            fragmentManager.unregisterFragmentLifecycleCallbacks(this);
                        }
                    }
                }, false);
        this.value = value;
    }

    public T get() {
        return value;
    }
}

File: common/src/main/java/com/aykuttasil/common/util/AutoClearedValue.kt
package com.aykuttasil.common.util

import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager

/**
 * A value holder that automatically clears the reference if the Fragment's view is destroyed.
 *
 * @param <T>
</T> */

class AutoClearedValue<T>(fragment: Fragment, private var value: T?) {

  init {
    val fragmentManager = fragment.parentFragmentManager
    fragmentManager.registerFragmentLifecycleCallbacks(
      object : FragmentManager.FragmentLifecycleCallbacks() {
        override fun onFragmentViewDestroyed(fm: FragmentManager, f: Fragment) {
          if (f === fragment) {
            this@AutoClearedValue.value = null
            fragmentManager.unregisterFragmentLifecycleCallbacks(this)
          }
        }
      },
      false
    )
  }

  fun get(): T? {
    return value
  }
}
